### ISSUE TITLE
cgroups should not be checked if substate is exited

### ISSUE THREAD

**Issue by JensErat** [prometheus-community/systemd_exporter#33]
## The Context

As described in #12, substates are often more describing. For example, we're using cloud-init, which has a unit of `Type=oneshot`. Those run a single time, and afterwards have `ActiveState=active` but `SubState=exited`.

Example unit:

```
$ systemctl status cloud-final
● cloud-final.service - Execute cloud user/final scripts
   Loaded: loaded (/lib/systemd/system/cloud-final.service; enabled; vendor preset: enabled)
   Active: active (exited) since Thu 2020-05-07 07:03:55 UTC; 1 weeks 6 days ago
  Process: 1118 ExecStart=/usr/bin/cloud-init modules --mode=final (code=exited, status=0/SUCCESS)
 Main PID: 1118 (code=exited, status=0/SUCCESS)
    Tasks: 0
   Memory: 544.9M
   CGroup: /system.slice/cloud-final.service
```

## The Issue

systemd_exporter tries to scrape cgroup information for this unit. From reading the code, I see this switch/case statement must be extended to exclude some substates (or probably better: include only really active ones, #12 has a more comprehensive list that can be used to derive): https://github.com/povilasv/systemd_exporter/blob/master/systemd/systemd.go#L578

**Comment by povilasv** [comment]
Thanks for reporting.

**Comment by ravi-arnan** [comment]
This looks obsolete on current `main`. The code it points at no longer exists.

The report targets the cgroup scraping at `systemd/systemd.go#L578` in the
povilasv tree. Since the rewrite there is no cgroup collection left in this
exporter at all:

```
$ grep -rn "cgroup\|ControlGroup" --include=*.go .
(no matches)
```

`collectUnit()` now dispatches to unit-state, unit-time, service metainfo,
start time, restart count, tasks, IP accounting, mount, timer and socket
collectors, and none of them reads `ControlGroup` or walks a cgroup hierarchy.

The specific `Type=oneshot` case from the report also behaves correctly now.
Checked against a live systemd 255 on Linux 6.17.0-35, on a unit in exactly the
reported state:

```
$ systemctl show alsa-restore.service -p ActiveState -p SubState -p TasksCurrent -p ControlGroup
ControlGroup=
TasksCurrent=[not set]
ActiveState=active
SubState=exited
```

`TasksCurrent` comes back as unset, which is `MaxUint64` over D-Bus, and
`collectServiceTasksMetrics()` already skips it:

```go
// Don't set if tasksCurrent if dbus reports MaxUint64.
if currentCount != math.MaxUint64 {
```

So no bogus `systemd_unit_tasks_current` is emitted for active(exited) units,
and there is no cgroup lookup to fail.

`collectServiceStartTimeMetrics()` does still read `ActiveEnterTimestamp` for
these units, but that is correct: an active(exited) oneshot really does have an
activation timestamp, and the value is real
(`ActiveEnterTimestamp=Tue 2026-08-18 14:34:19 WITA` for the unit above).

Unless someone is still seeing errors from this on a recent build, I would
suggest closing. Happy to be wrong if there is a cgroup path I have missed.

**Comment by JensErat** [comment]
I agree with your observation, and will close this issue. Thank you for the analysis.
